FWCore/Framework/src/ProductResolvers.cc

```diff
@@ -516,6 +516,17 @@ namespace edm {
     return Resolution(nullptr);
   }
 
+  void TransformingProductResolver::putProduct(std::unique_ptr<WrapperBase> edp) const {
+    // Override putProduct() to not set the resolver status to
+    // ResolveFailed when the Event::commit_() checks which produced
+    // products were actually produced and which not, because the
+    // transforming products are never produced by time of commit_()
+    // by construction.
+    if (edp) {
```

Comment:
Should the putProduct call always get a null wrapper? Or does the later call also call putProduct?

Reply:
The `putProduct()` call from `Event::commit_()`
https://github.com/cms-sw/cmssw/blob/ff2213a4904f42937039cad45098572009abdc40/FWCore/Framework/src/Event.cc#L170-L175
will always have a null wrapper. The Transformer itself calls `putProduct()` via `EventPrincipal::put()`
https://github.com/cms-sw/cmssw/blob/ff2213a4904f42937039cad45098572009abdc40/FWCore/Framework/src/EventPrincipal.cc#L209
https://github.com/cms-sw/cmssw/blob/ff2213a4904f42937039cad45098572009abdc40/FWCore/Framework/src/EventForTransformer.cc#L22-L26

Written that, I see now an alternative to customizing `putProduct()` that would be to customize (in some way) `productResolved()` to return `true` for Transformer. The `productResolved()` seems to be called only in `{Event,LuminosityBlock,Run,ProcessBlock}::commit_()`, but on a first thought "lying" in `productResolved()` sounds conceptually worse than leaving the status unaltered in `putProduct()`.